### PR TITLE
Fix warnings for size, ..density.., and expand_scale from ggplot2 0.3.3/0.3.4

### LINKED ISSUE #78: ggplot2 3.4.0 warning for size vs linewidth
The `size` argument for the geom_vline internals of ridges is throwing a warning now that ggplot2 has a linewidth argument and deprecated size.

``` r
library(ggplot2)
library(ggridges)

mtcars |>
  ggplot() +
  geom_density_ridges(
    aes(x = mpg, y = cyl, group = cyl),
    quantile_lines = TRUE,
    quantiles = 2,
    size = 0.5
  )
#> Picking joint bandwidth of 1.38
#> Warning: Using the `size` aesthietic with geom_segment was deprecated in ggplot2 3.4.0.
#> ℹ Please use the `linewidth` aesthetic instead.
```

``` r

sessioninfo::session_info(pkgs = "loaded")
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.0 (2022-04-22)
#>  os       macOS Monterey 12.6
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/Chicago
#>  date     2022-11-17
#>  pandoc   2.19.2 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date (UTC) lib source
#>  assertthat    0.2.1       2019-03-21 [1] CRAN (R 4.2.0)
#>  cli           3.4.1       2022-09-23 [1] CRAN (R 4.2.0)
#>  colorspace    2.0-3       2022-02-21 [1] CRAN (R 4.2.0)
#>  DBI           1.1.3       2022-06-18 [1] CRAN (R 4.2.0)
#>  digest        0.6.30      2022-10-18 [1] CRAN (R 4.2.0)
#>  dplyr         1.0.10      2022-09-01 [1] CRAN (R 4.2.0)
#>  evaluate      0.16        2022-08-09 [1] CRAN (R 4.2.0)
#>  fansi         1.0.3       2022-03-24 [1] CRAN (R 4.2.0)
#>  farver        2.1.1       2022-07-06 [1] CRAN (R 4.2.0)
#>  fastmap       1.1.0       2021-01-25 [1] CRAN (R 4.2.0)
#>  fs            1.5.2       2021-12-08 [1] CRAN (R 4.2.0)
#>  generics      0.1.3       2022-07-05 [1] CRAN (R 4.2.0)
#>  ggplot2     * 3.4.0       2022-11-04 [1] CRAN (R 4.2.0)
#>  ggridges    * 0.5.4       2022-11-17 [1] Github (wilkelab/ggridges@add4f72)
#>  glue          1.6.2       2022-02-24 [1] CRAN (R 4.2.0)
#>  gtable        0.3.1       2022-09-01 [1] CRAN (R 4.2.0)
#>  highr         0.9         2021-04-16 [1] CRAN (R 4.2.0)
#>  htmltools     0.5.3       2022-07-18 [1] CRAN (R 4.2.0)
#>  knitr         1.40        2022-08-24 [1] CRAN (R 4.2.0)
#>  labeling      0.4.2       2020-10-20 [1] CRAN (R 4.2.0)
#>  lifecycle     1.0.3       2022-10-07 [1] CRAN (R 4.2.0)
#>  magrittr      2.0.3       2022-03-30 [1] CRAN (R 4.2.0)
#>  munsell       0.5.0       2018-06-12 [1] CRAN (R 4.2.0)
#>  pillar        1.8.1       2022-08-19 [1] CRAN (R 4.2.0)
#>  pkgconfig     2.0.3       2019-09-22 [1] CRAN (R 4.2.0)
#>  purrr         0.3.5       2022-10-06 [1] CRAN (R 4.2.0)
#>  R.cache       0.15.0      2021-04-30 [1] CRAN (R 4.2.0)
#>  R.methodsS3   1.8.1       2020-08-26 [1] CRAN (R 4.2.0)
#>  R.oo          1.24.0      2020-08-26 [1] CRAN (R 4.2.0)
#>  R.utils       2.11.0      2021-09-26 [1] CRAN (R 4.2.0)
#>  R6            2.5.1       2021-08-19 [1] CRAN (R 4.2.0)
#>  reprex        2.0.1       2021-08-05 [1] CRAN (R 4.2.0)
#>  rlang         1.0.6       2022-09-24 [1] CRAN (R 4.2.0)
#>  rmarkdown     2.14        2022-04-25 [1] CRAN (R 4.2.0)
#>  rstudioapi    0.14.0-9000 2022-09-01 [1] local
#>  scales        1.2.1       2022-08-20 [1] CRAN (R 4.2.0)
#>  sessioninfo   1.2.2       2021-12-06 [1] CRAN (R 4.2.0)
#>  stringi       1.7.8       2022-07-11 [1] CRAN (R 4.2.0)
#>  stringr       1.4.1       2022-08-20 [1] CRAN (R 4.2.0)
#>  styler        1.8.1       2022-11-07 [1] CRAN (R 4.2.0)
#>  tibble        3.1.8       2022-07-22 [1] CRAN (R 4.2.0)
#>  tidyselect    1.2.0.9000  2022-11-16 [1] Github (r-lib/tidyselect@b449033)
#>  utf8          1.2.2       2021-07-24 [1] CRAN (R 4.2.0)
#>  vctrs         0.5.1       2022-11-16 [1] CRAN (R 4.2.0)
#>  withr         2.5.0       2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun          0.33        2022-09-12 [1] CRAN (R 4.2.0)
#>  yaml          2.3.5       2022-02-21 [1] CRAN (R 4.2.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

<sup>Created on 2022-11-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

```r
Warning message:
Using the `size` aesthietic with geom_segment
was deprecated in ggplot2 3.4.0.
ℹ Please use the `linewidth` aesthetic instead.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to
see where this warning was generated. 
[15:45 💻 @main]> lifecycle::last_lifecycle_warnings()
[[1]]
<warning/lifecycle_warning_deprecated>
Warning:
Using the `size` aesthietic with geom_segment
was deprecated in ggplot2 3.4.0.
ℹ Please use the `linewidth` aesthetic instead.
---
Backtrace:
  1. base (local) `<fn>`(x)
  2. ggplot2:::print.ggplot(x)
  4. ggplot2:::ggplot_gtable.ggplot_built(data)
  5. ggplot2:::by_layer(...)
 12. ggplot2 (local) f(l = layers[[i]], d = data[[i]])
 13. l$draw_geom(d, layout)
 14. ggplot2 (local) draw_geom(..., self = self)
 15. self$geom$draw_layer(...)
 16. ggplot2 (local) draw_layer(..., self = self)
 17. base::lapply(...)
 18. ggplot2 (local) FUN(X[[i]], ...)
 20. self$draw_panel(data, panel_params, coord, na.rm = FALSE)
 21. ggridges (local) draw_panel(..., self = self)
 22. base::lapply(...)
 23. ggridges (local) FUN(X[[i]], ...)
 24. self$draw_group(group, panel_params, coord, ...)
 25. ggridges (local) draw_group(..., self = self)
 26. self$make_vline_grob(data_list[["vline"]], panel_params, coord)
 27. ggridges (local) make_vline_grob(...)
 28. ggplot2::GeomSegment$draw_panel(data, panel_params, coord)
 29. ggplot2 (local) draw_panel(..., self = self)
 30. ggplot2:::check_linewidth(data, snake_class(self))
 31. ggplot2:::deprecate_soft0(...)
```

### PR DESCRIPTION
**Pull request by jthomasmock**
ggridges 0.5.5
----------------------------------------------------------------
- Replace `ggplot2::expand_scale()` with `ggplot2::expansion()` in vignettes to avoid deprecation warnings in [ggplot2 0.3.3](https://ggplot2.tidyverse.org/news/index.html#minor-improvements-and-bug-fixes-3-3-0)
- Replace `size` argument in line functions with `linewidth` to avoid deprecation warnings in [ggplot2 0.3.4](https://ggplot2.tidyverse.org/news/index.html#breaking-changes-3-4-0)
- Replace use of `..density..` with `ggplot2::after_stat(density)` to avoid deprecation warnings in [ggplot2 0.3.4](https://ggplot2.tidyverse.org/news/index.html#breaking-changes-3-4-0)

Closes #78

### CODE REVIEW

**Comment by clauswilke** [comment]
Super, thanks! Do you want to acknowledge your contribution in the changelog by adding `(jthomasmock, #78)`?

**Comment by jthomasmock** [comment]
> Super, thanks! Do you want to acknowledge your contribution in the changelog by adding `(jthomasmock, #78)`?

Done!

**Comment by clauswilke** [comment]
There's one more issue, and it's a bit of a tricky one (maybe). I invented alternative aesthetics so people could customize the points and vertical lines separately. See for example here: https://github.com/wilkelab/ggridges/blob/add4f72193a4a0142d0ee90724f5427c037d7911/R/geoms.R#L110-L116

In the code, I'm using either the point/line specific one or the general one, whichever is provided:
https://github.com/wilkelab/ggridges/blob/add4f72193a4a0142d0ee90724f5427c037d7911/R/geoms.R#L167

It seems wrong to me to only change `size` to `linewidth` and not change `vline_size` also to `vline_linewidth`. However, with the latter change, it may break some people's code, as there's no fallback. So what do we do? Maybe just breaking the code is better than carrying fallback code forward for all eternity.

**Comment by jthomasmock** [comment]
I saw that, but in the moment decided to leave alone since it may be much more disruptive (and I didn't know the code base that well).

I guess that the size -> linewidth change is just as disruptive, since the size option was removed...through that lens I'd be in favor of "this is the breaking change patch" rather than keeping linesize around for longer.

AFAIK if we did rename linesize, if a user used the old argument it would not error but rather throw a warning that the aesthetic is ignored. That feels ok in that we're not causing errors but a warning and clarifying in changelog that the option is renamed.

**Comment by clauswilke** [comment]
Yes, no code will break. Just some figures will no longer look as expected. And this applies only to the extent that anybody even uses this feature. I don't know if that's the case. It's rather obscure, so may not be a big deal to just break it.

**Comment by jthomasmock** [comment]
> Yes, no code will break. Just some figures will no longer look as expected. And this applies only to the extent that anybody even uses this feature. I don't know if that's the case. It's rather obscure, so may not be a big deal to just break it.

I'm of the opinion that changing linewidth with aes() is a less-used feature - I can try to add to the PR a specific warning if `line_size` arg is used? That would be a more explicit warning but would also require you to remove that warning at some point in future.

**Comment by clauswilke** [comment]
Note that whether we use an aesthetic inside of `aes()` or rather as a parameter directly in the geom doesn't make any difference to how it's handled downstream at drawing time.

The more I think about it, the more I believe we should just rename `vline_size` to `vline_linewidth` throughout. The warning would be great if it's not too much work, as it would tell people they need to change their code rather than just silently producing a changed plot. I don't think the warning has to ever be removed if it says something like "`vline_size` has been deprecated; please use `vline_linewidth` instead."

**Comment by jthomasmock** [comment]
Ok, I'll try and update this PR.

Plan:

Convert all instance of `vline_size` to `vline_linewidth` and take a pass at detecting the `vline_size` argument for a special warning. I am confident in the first half, and less so in the second but will take a shot.

**Comment by jthomasmock** [comment]
I _think_ I can use logic similar to this:

https://github.com/tidyverse/ggplot2/blob/63125db184b6e7d5516007574b5d15acbfd8f470/R/geom-.r#L119-L122

**Comment by clauswilke** [comment]
I think this is the right logic for the check:
https://github.com/tidyverse/ggplot2/blob/63125db184b6e7d5516007574b5d15acbfd8f470/R/geom-.r#L241-L247

**Comment by jthomasmock** [comment]
Unfortunately, since I'm _replacing_ and not actually deprecating the arguments are simply not valid, and the warnings are intercepted by missing args warnings. I could add the arguments back alongside the new `vline_width`, but that also further complicates the code.

With the logic as outlined above in the `ggplot2` examples:

```
# aesthetic ignored
Warning message:
In geom_density_ridges(aes(vline_size = cyl), quantile_lines = TRUE) :
  Ignoring unknown aesthetics: vline_size

# parameter ignored
Warning message:
In geom_density_ridges(quantile_lines = TRUE, vline_size = 1) :
  Ignoring unknown parameters: `vline_size`
```

While that's not our "intended" message - it does at least tell the user that their parameter/aesthetic is no longer valid - leading them to check the arguments/aesthetics. It's the same warning for `vline_size` and `size` now, where `linewidth` or `vline_width` work silently as intended.

``` r
library(ggplot2)
library(ggridges) #<- local dev version

base_plt <- mtcars |> 
  ggplot(aes(x = mpg, y = factor(cyl))) 

# not a valid aesthetic
base_plt +
  geom_density_ridges(aes(vline_size = cyl), quantile_lines = TRUE)
#> Warning in geom_density_ridges(aes(vline_size = cyl), quantile_lines = TRUE):
#> Ignoring unknown aesthetics: vline_size
#> Picking joint bandwidth of 1.38
```



``` r

# not a valid parameter
base_plt +
  geom_density_ridges(quantile_lines = TRUE, vline_size = 1)
#> Warning in geom_density_ridges(quantile_lines = TRUE, vline_size = 1): Ignoring
#> unknown parameters: `vline_size`
#> Picking joint bandwidth of 1.38
```



``` r

# swap and works
base_plt +
  geom_density_ridges(aes(vline_width = cyl), quantile_lines = TRUE)
#> Picking joint bandwidth of 1.38
```



``` r

# swap and works
base_plt +
  geom_density_ridges(quantile_lines = TRUE, vline_width = 3)
#> Picking joint bandwidth of 1.38
```



<sup>Created on 2022-11-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by clauswilke** [comment]
I think you can keep the old variant (`vline_size`) alive by listing it as `optional_aes` here:
https://github.com/wilkelab/ggridges/blob/add4f72193a4a0142d0ee90724f5427c037d7911/R/geoms.R#L121

Then you can implement the check as in ggplot.

If that doesn't work, I'd probably give `vline_size` a default value (in `default_aes`), even if it's not used anymore. I think that's better than throwing the seemingly random error that `vline_size` is an unknown parameter.

**Comment by jthomasmock** [comment]
I have kept the `vline_size` arg as NULL and provided a deprecation warning.

``` r
library(ggplot2)
library(ggridges) #<- local dev version

base_plt <- mtcars |> 
  ggplot(aes(x = mpg, y = factor(cyl))) 

# not a valid aesthetic
base_plt +
  geom_density_ridges(aes(vline_size = cyl), quantile_lines = TRUE)
#> Picking joint bandwidth of 1.38
#> Warning: Use of the `vline_size` or `size` aesthetics are deprecated, please use
#> `linewidth` instead of `size` and `vline_width` instead of `vline_size`.
```



<sup>Created on 2022-11-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by clauswilke** [comment]
Thanks for all your effort! We're almost there, but there's one small issue remaining. The aesthetic `vline_width` only applies to vertical lines, whereas `linewidth` applies to all lines. In your last reprex, when setting `vline_size` you can see how it changes all lines, not just the vertical lines, whereas in earlier reprexes   you made that wasn't the case. The issue is that in your check function you overwrite `linewidth` with `vline_size`. I think we need two separate checks, one for `size` (which gets copied to `linewidth`) and one for `vline_size` (which gets copied to `vline_width`).

File: R/utils.R [hunk 1]
```diff
@@ -36,3 +36,20 @@ reduce <- function(.x, .f, ..., .init) {
 
   v
 }
+
+check_vline_size <- function(data) {
+  if (is.null(data$linewidth) && !is.null(data$vline_size)) {
```
Comment (clauswilke) [inline R/utils.R]: This should be `is.null(data$vline_width) && !is.null(data$vline_size)` and then we need a second check for `linewidth` and `size`, respectively.

File: R/utils.R [hunk 2]
```diff
@@ -36,3 +36,20 @@ reduce <- function(.x, .f, ..., .init) {
...
+    warning("  Use of the `vline_size` or `size` aesthetics are deprecated, please use `linewidth` instead of `size` and `vline_width` instead of `vline_size`.", call. = FALSE)
+    data$linewidth <- data$vline_size
+  }
+  data
+}
+
+check_vline_size_param <- function(params) {
+  if ("vline_size" %in% names(params)) {
```
Comment (clauswilke) [inline R/utils.R:57]: Do we need this also for `size`?

File: R/utils.R [hunk 3]
```diff
@@ -36,3 +36,20 @@ reduce <- function(.x, .f, ..., .init) {
...
+    warning("  Use of the `vline_size` or `size` aesthetics are deprecated, please use `linewidth` instead of `size` and `vline_width` instead of `vline_size`.", call. = FALSE)
+    params$vline_width <- params$vline_size
```
Comment (clauswilke) [inline R/utils.R:59]: This is correct: copy `vline_size` over to `vline_width`.

File: R/utils.R (hunk 1 above)
Comment (jthomasmock) [reply R/utils.R]: Added

File: R/utils.R (hunk 2 above)
Comment (jthomasmock) [reply R/utils.R:57]: Added

**Comment by clauswilke** [comment]
Super, thanks! If the CI runs without issues I'll merge.

**Comment by clauswilke** [comment]
Great: `r-lib/actions/setup-r@v1 is deprecated. Please update your workflow to use the 'v2' version.`

Let me know if you're Ok to take that on also. It should probably be a separate PR.

**Comment by clauswilke** [comment]
Ok, only the windows check fails, with the error message I posted, so no reason to assume anything is wrong with this PR. I'll go ahead and merge. Thanks again!

**Pull request merged**